### DISCUSSION TITLE
Wifi protocol

### DISCUSSION THREAD

**Discussion by iloveicedgreentea** [jgyates/genmon#1157] [Ideas]
Hey

I was wondering if you explored Generac's wifi protocols? Instead of using RS232 I was hoping to set up a wifi proxy or emulate their cloud API locally. I don't see any open ports or anything... so I wanted to see if this was explored already.

No open ports means it only initiates outbound connections (of course they would do that). So its impossible to do this via wifi without making a MITM proxy appliance

**Comment by madisdev** [comment]
Not sure why that is necessary, If genmon has direct access to the generator controller why would you want to get data from another avenue?

**Reply by iloveicedgreentea** [reply]
Because its 100% more accessible to do it with no additional hardware

**Reply by madisdev** [reply]
I mean to each their own, I didnt find it that difficult to setup, hardware is cheap for this project. The only real factor is time, and fortunately for us the majority of the work was done by @jgyates.

What genmon does in their wifi controller is a closed system. The effort to replicate what they are doing doesn't seem like a couple hour project and IMHO I am not sure what benefit we would get out of it. I highly doubt they send all the datapoints that are currently available through genmon interface.

**Reply by jgyates** [reply]
The mobile link does not support any locally callable API that I am aware of. The firmware calls a predefined URL to submit data that uses encryption. This could be intercepted via a proxy, but I have not explored this at all. This is how it is able to traverse local firewalls.

This project was initially started before the wifi based option and has expaned to support other modbus based generator controllers so I have not fully looked into other methods. The Mobile Link wifi module does communicate with the controller via RS-485 using a private protocol (not modbus).

**Reply by iloveicedgreentea** [reply]
thanks that is helpful. I am going to explore the feasibility of a simple proxy to see what kind of data I can get locally

**Reply by jgyates** [reply]
Sounds good, let me know if you find anything useful. I could possible add this to the project if you like.

**Reply by iloveicedgreentea** [reply]
The device uses cert pinning and will not accept self signed/invalid certs

It polls device-prod.generaccloud.com, prod.otafu.generac.com, and  time-prod.generaccloud.com every 20-45 min

Unfortunately this means MITM is not possible without firmware modification which I don't want to attempt. So unfortunately using an external device is going to be the only viable option unless someone reverse engineers the firmware.
